### PR TITLE
Allow for large Models.json file

### PR DESCRIPTION
**Pull request by SimonWalton-HiFi**
Fogbugz: https://highfidelity.fogbugz.com/f/cases/17179/

Qt's QJsonDocument has some built-in 128-MiB limits; see attached issue for details.
This PR works around the issue for reading by building the QVariant one json entity at a time, so the limit only applies to each entity. It works around the issue for writing by also writing json one entity at a time, overriding the stringify for the EntityProperties to js.

It also reduces some of the multiple reading by the entity server by adding a cache of the entire file.

### CODE REVIEW

File: libraries/octree/src/OctreeEntitiesFileParser.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+#include <cctype>
+#include <QUuid>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+#include "OctreeEntitiesFileParser.h"
+
+using std::string;
+
+OctreeEntitiesFileParser::OctreeEntitiesFileParser() {
```
Comment (ZappoMan) [inline libraries/octree/src/OctreeEntitiesFileParser.cpp]: is this really an "octree" file parser? or is it a generic JSON file parser?

Comment (SimonWalton-HiFi) [reply libraries/octree/src/OctreeEntitiesFileParser.cpp]: It is not generic - it's specific to the Models.json format that we save the entities to.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14467/HighFidelity-Beta-PR14294-be279f0.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14467/HighFidelity-Beta-PR14294-be279f0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14467/HighFidelity-Beta-Interface-PR14294-be279f0.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14487/HighFidelity-Beta-PR14294-6f6c92e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14487/HighFidelity-Beta-PR14294-6f6c92e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14487/HighFidelity-Beta-Interface-PR14294-6f6c92e.dmg

**Comment by SimonWalton-HiFi** [comment]
## Test Plan
The large domain backup is [here](https://highfidelity.fogbugz.com/default.asp?ixAttachment=7685&pg=pgDownload&pgType=pgFile&sFilename=autobackup-thirty_day_rolling-2018-07-13_22-49-19.content.zip) (also available from the Fogbugz). 

- Run a sandbox and client from the PR.
- Unzip the large domain content locally.
- Upload the models.json.gz file via the domain server (http://localhost:40100/content/#upload_content_group -> Choose File -> Upload Content)
- Observe the painting domain from the client. Add several paint strokes.
- Shutdown the sandbox then restart it
- Reconnect to the painting domain

With the PR you should be able to restart domains that are too large for master. As noted in Fogbugz, PR 16341 will reduce the size of saved domains, but you still should be able to go over the limit in master but not in the PR. The domain is saved as a gzipped JSON file in (on Windows) `AppData/Roaming/High Fidelity - PR.../assignment-client/entities/models.json.gz`.

- Compare saving and loading times for the models file file from the entity server. These can be observed from the entity-server logs - for reading the log line "loading Octrees from file: ..." should be followed by "DONE loading Octrees from file..." when read and loaded. For writing the log line "Saving Octree data to: ..." should be followed by "DONE persisting Octree data to ..." when writing completes.

There should be no significant regressions in the times for various-sized domains.

**Comment by birarda** [comment]
We'll need to double check performance testing here - the repeated entity server persists can't be significantly slower.

File: libraries/entities/src/RecurseOctreeToJSONOperator.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "RecurseOctreeToJSONOperator.h"
+#include "EntityItemProperties.h"
+
+RecurseOctreeToJSONOperator::RecurseOctreeToJSONOperator(const OctreeElementPointer&, QScriptEngine* engine,
+    QString jsonPrefix /* = QString() */, bool skipDefaults /* = true */)
+    : _engine(engine)
+    , _json(jsonPrefix)
+    , _skipDefaults(skipDefaults)
+{
+    _toStringMethod = _engine->evaluate("(function() { return JSON.stringify(this, null, '    ') })");
```
Comment (birarda) [inline libraries/entities/src/RecurseOctreeToJSONOperator.cpp]: We should check if using this or a conversion from QScriptValue to QVariant to JSON is faster (I'm guessing this, but I think it's worth checking)

File: libraries/octree/src/OctreeDataUtils.cpp [hunk 3]
```diff
@@ -49,19 +56,30 @@ bool OctreeUtils::RawOctreeData::readOctreeDataInfoFromJSON(QJsonObject root) {
     return true;
 }
 
+bool OctreeUtils::RawOctreeData::readOctreeDataInfoFromMap(QVariantMap map) {
```
Comment (birarda) [inline libraries/octree/src/OctreeDataUtils.cpp]: I'm guessing copy-on-write makes this not an issue but we should take a `const QVariantMap&` here?

File: libraries/octree/src/OctreeEntitiesFileParser.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+}
```
Comment (birarda) [inline libraries/octree/src/OctreeEntitiesFileParser.cpp]: Seems like this matches the default and should be removable

File: libraries/octree/src/OctreeEntitiesFileParser.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+            parsedEntities["DataVersion"] = dataVersionValue;
+            gotDataVersion = true;
+        } else if (key == "Entities") {
+            if (gotEntities) {
+                _errorString = "Duplicate Entities entries";
+                return false;
+            }
+
+            QVariantList entitiesValue;
+            if (!readEntitiesArray(entitiesValue)) {
```
Comment (birarda) [inline libraries/octree/src/OctreeEntitiesFileParser.cpp]: Can we make `parsedEntities["Entities"]` start with an empty QVariantList and pass it to `readEntitiesArray` so we avoid the copy on line 89?

File: libraries/octree/src/Octree.cpp [hunk 6]
```diff
@@ -878,17 +882,75 @@ bool Octree::toJSONDocument(QJsonDocument* doc, const OctreeElementPointer& elem
...
+    OctreeElementPointer top;
+    if (element) {
+        top = element;
+    } else {
+        top = _rootElement;
+    }
+
+    jsonString += QString(R"({
+  "DataVersion": %1,
+  "Entities": [)").arg(_persistDataVersion);
```
[the remark this answers is not in the document]
Comment (SimonWalton-HiFi) [reply libraries/octree/src/Octree.cpp]: The library conversion to JSON (called via line 911) unfortunately doesn't do nested indenting, only allowing a fixed indent. Using two spaces in these cases was an attempt to get reasonable looking JSON files. I can take another look.

File: libraries/octree/src/OctreeEntitiesFileParser.cpp (hunk 5 above)
Comment (SimonWalton-HiFi) [reply libraries/octree/src/OctreeEntitiesFileParser.cpp]: I used std::move() instead. I think the compiler was calling move assignment anyway but this makes it explicit.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14991/HighFidelity-Beta-PR14294-d183968.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14991/HighFidelity-Beta-PR14294-d183968.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14991/HighFidelity-Beta-Interface-PR14294-d183968.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14991/HighFidelity-Beta-PR14294-d183968.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14991/HighFidelity-Beta-Interface-PR14294-d183968.exe

**Comment by SimonWalton-HiFi** [comment]
The domain server from this PR currently won't upload an entities.json file. The domain server parses and rebuilds the entities file (changing some fields) before passing to the entity server. The OctreeDataUtils in this PR won't rewrite the original entities. I will commit a fix later.

**Comment by SimonWalton-HiFi** [comment]
The domain server issue should be fixed now.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15055/HighFidelity-Beta-PR14294-6ee837d.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15055/HighFidelity-Beta-PR14294-6ee837d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15055/HighFidelity-Beta-Interface-PR14294-6ee837d.exe

**Comment by qligmaloney** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15071/HighFidelity-Beta-PR14294-6ee837d.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15071/HighFidelity-Beta-PR14294-6ee837d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15071/HighFidelity-Beta-Interface-PR14294-6ee837d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15078/HighFidelity-Beta-PR14294-df2bc6b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15078/HighFidelity-Beta-Interface-PR14294-df2bc6b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15078/HighFidelity-Beta-PR14294-df2bc6b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15078/HighFidelity-Beta-Interface-PR14294-df2bc6b.exe

**Comment by qlipdahlin** [comment]
Currently Testing

File: libraries/octree/src/Octree.cpp [hunk 7]
```diff
@@ -878,17 +882,75 @@ bool Octree::toJSONDocument(QJsonDocument* doc, const OctreeElementPointer& elem
...
+    if (doGzip) {
+        if (!gzip(jsonString.toUtf8(), *data, -1)) {
+            qCritical("Unable to gzip data while saving to json.");
+            return false;
+        }
+    } else {
+        *data = jsonString.toUtf8();
+    }
+
+#else
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply libraries/octree/src/Octree.cpp]: agreed

File: libraries/octree/src/Octree.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (birarda) [reply libraries/octree/src/Octree.cpp]: Could we put this (and the one below) on one line with explicit new line characters? This is really hard to glance at.

File: libraries/octree/src/OctreeDataUtils.cpp [hunk 8]
```diff
@@ -55,40 +35,46 @@ bool OctreeUtils::RawOctreeData::readOctreeDataInfoFromData(QByteArray data) {
...
 
-    writeSubclassData(obj);
+    writeSubclassData(jsonString);
 
-    QJsonDocument doc;
-    doc.setObject(obj);
+    jsonString += QString(R"(,
+  "Id": "%1",
+  "Version": %2
+})").arg(id.toString()).arg(version);
```
Comment (birarda) [inline libraries/octree/src/OctreeDataUtils.cpp]: Same one line comment here

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15098/HighFidelity-Beta-PR14294-a65466f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15098/HighFidelity-Beta-PR14294-a65466f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15098/HighFidelity-Beta-Interface-PR14294-a65466f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15098/HighFidelity-Beta-PR14294-a65466f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15098/HighFidelity-Beta-Interface-PR14294-a65466f.exe

**Comment by SimonWalton-HiFi** [comment]
I did some timing tests of this branch. With both the zombies scene (2.1 MB uncompressed) and dev-mobile (800 KB) time to read and save JSON was < 1 s (ie. below the resolution of the timestamps).

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15127/HighFidelity-Beta-PR14294-82e21e8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15127/HighFidelity-Beta-PR14294-82e21e8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15127/HighFidelity-Beta-Interface-PR14294-82e21e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15127/HighFidelity-Beta-PR14294-82e21e8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15127/HighFidelity-Beta-Interface-PR14294-82e21e8.exe

**Comment by SimonWalton-HiFi** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15135/HighFidelity-Beta-PR14294-452c212.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15135/HighFidelity-Beta-Interface-PR14294-452c212.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15135/HighFidelity-Beta-PR14294-452c212.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15135/HighFidelity-Beta-Interface-PR14294-452c212.exe

**Comment by qlipdahlin** [comment]
Currently Testing

**Comment by qlipdahlin** [comment]
@conklin94122 PASS

Serverless domain loads without issue now.

Retested feature and it still works.

There was an issue with some interface menu's not closing or doing as they should, but is present on Master.

PR Passes.

**Comment by SimonWalton-HiFi** [comment]
I've checked that, with the update from master, json vec3s are no longer written redundantly as six components.

**Pull request merged**